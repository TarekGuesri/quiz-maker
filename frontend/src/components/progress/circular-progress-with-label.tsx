import React from "react";
import CircularProgress, {
  CircularProgressProps,
} from "@mui/material/CircularProgress";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";

export const CircularProgressWithLabel = (
  props: CircularProgressProps & { value: number },
) => {
  return (
    <Box sx={{ position: "relative", display: "inline-flex" }}>
      <CircularProgress
        variant="determinate"
        size={160}
        sx={{
          color: (theme) =>
            theme.palette.mode === "light"
              ? theme.palette.primary.light
              : theme.palette.primary.dark,
        }}
        {...props}
        color="secondary"
        value={100}
      />
      <CircularProgress
        disableShrink
        sx={{
          color: (theme) =>
            theme.palette.mode === "light"
              ? theme.palette.primary.dark
              : theme.palette.primary.light,
          animationDuration: "550ms",
          position: "absolute",
          left: 0,
        }}
        variant="determinate"
        size={160}
        {...props}
      />
      <Box
        sx={{
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
          position: "absolute",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        <Typography
          variant="h4"
          component="div"
          color="text.secondary"
        >{`${Math.round(props.value)}%`}</Typography>
      </Box>
    </Box>
  );
};
