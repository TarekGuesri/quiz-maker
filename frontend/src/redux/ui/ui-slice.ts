import { createSlice } from "@reduxjs/toolkit";
import { UIState } from "src/types";

const initialState: UIState = {
  darkMode: localStorage.getItem("darkMode") === "on",
};

export const uiSlice = createSlice({
  name: "ui",
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    switchDarkMode: (state) => {
      state.darkMode = !state.darkMode;

      localStorage.setItem("darkMode", state.darkMode ? "on" : "off");
    },
  },
});

export const { switchDarkMode } = uiSlice.actions;

export default uiSlice.reducer;
