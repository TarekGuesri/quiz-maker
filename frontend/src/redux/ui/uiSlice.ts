import { createSlice } from '@reduxjs/toolkit';

export interface UIState {
  darkMode: boolean;
}

const initialState: UIState = {
  darkMode: true,
};

export const uiSlice = createSlice({
  name: 'ui',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {},
});

export default uiSlice.reducer;
