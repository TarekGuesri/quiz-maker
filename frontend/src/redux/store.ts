import { Action, ThunkAction, configureStore } from "@reduxjs/toolkit";

import ui from "./ui/ui-slice";

export const store = configureStore({
  reducer: {
    ui,
  },
  devTools: true,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
