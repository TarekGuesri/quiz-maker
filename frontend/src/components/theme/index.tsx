import React, { FC } from 'react';
import CssBaseline from '@mui/material/CssBaseline';
import PropTypes from 'prop-types';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useAppSelector } from 'src/redux/hooks';
import varelaRound from 'src/assets/fonts/VarelaRound-Regular.ttf';

export const Theme: FC = (props) => {
  const { darkMode } = useAppSelector((state) => state.ui);

  const theme = createTheme({
    palette: {
      mode: darkMode ? 'dark' : 'light',
      background: {
        default: darkMode ? '#333' : '#f7f7f7',
        paper: darkMode ? '#333' : '#fbfbfb',
      },
      primary: {
        main: '#8e49de',
        // contrastText:
      },
      secondary: {
        main: '#f9de4a',
      },
      contrastThreshold: 3,
      tonalOffset: 0.2,
    },
    typography: {
      fontFamily: 'Varela Round',
    },
    components: {
      MuiCssBaseline: {
        styleOverrides: `
          @font-face {
            font-family: 'Varela Round';
            font-style: normal;
            font-display: swap;
            font-weight: 400;
            src: url(${varelaRound}) format('truetype');
        `,
      },
    },
  });

  return (
    <ThemeProvider theme={theme}>
      {props.children}
      <CssBaseline />
    </ThemeProvider>
  );
};

Theme.propTypes = {
  children: PropTypes.node.isRequired,
};
