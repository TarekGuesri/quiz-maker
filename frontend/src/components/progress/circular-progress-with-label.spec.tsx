import React from "react";
import { cleanup, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { Theme } from "src/components/theme";
import { store } from "src/redux/store";
import { switchDarkMode } from "src/redux/ui/ui-slice";
import { CircularProgressWithLabel } from "./circular-progress-with-label";

describe("src/components/progress/circular-progress-with-label", () => {
  afterEach(() => cleanup());

  test("Render CircularProgressWithLabel component in light mode", () => {
    const renderedComponent = render(
      <Provider store={store}>
        <Theme>
          <CircularProgressWithLabel value={50} />
        </Theme>
      </Provider>,
    );

    expect(renderedComponent.container).toMatchSnapshot();
  });

  test("Render CircularProgressWithLabel component in dark mode", () => {
    const renderedComponent = render(
      <Provider store={store}>
        <Theme>
          <CircularProgressWithLabel value={50} />
        </Theme>
      </Provider>,
    );

    store.dispatch(switchDarkMode());

    expect(renderedComponent.container).toMatchSnapshot();
  });
});
