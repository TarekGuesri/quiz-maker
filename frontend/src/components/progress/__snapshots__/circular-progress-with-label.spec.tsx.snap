// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/progress/circular-progress-with-label Render CircularProgressWithLabel component in dark mode 1`] = `
<div>
  <div
    class="MuiBox-root css-rmz4vk"
  >
    <span
      aria-valuenow="100"
      class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorSecondary css-185cjip-MuiCircularProgress-root"
      role="progressbar"
      style="width: 160px; height: 160px; transform: rotate(-90deg);"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
          style="stroke-dasharray: 126.920; stroke-dashoffset: 0.000px;"
        />
      </svg>
    </span>
    <span
      aria-valuenow="50"
      class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-1kz5d1-MuiCircularProgress-root"
      role="progressbar"
      style="width: 160px; height: 160px; transform: rotate(-90deg);"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
          style="stroke-dasharray: 126.920; stroke-dashoffset: 63.460px;"
        />
      </svg>
    </span>
    <div
      class="MuiBox-root css-1o8dslu"
    >
      <div
        class="MuiTypography-root MuiTypography-h4 css-1tradvb-MuiTypography-root"
      >
        50%
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/progress/circular-progress-with-label Render CircularProgressWithLabel component in light mode 1`] = `
<div>
  <div
    class="MuiBox-root css-rmz4vk"
  >
    <span
      aria-valuenow="100"
      class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorSecondary css-5nap9f-MuiCircularProgress-root"
      role="progressbar"
      style="width: 160px; height: 160px; transform: rotate(-90deg);"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
          style="stroke-dasharray: 126.920; stroke-dashoffset: 0.000px;"
        />
      </svg>
    </span>
    <span
      aria-valuenow="50"
      class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-1pzb5l9-MuiCircularProgress-root"
      role="progressbar"
      style="width: 160px; height: 160px; transform: rotate(-90deg);"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
          style="stroke-dasharray: 126.920; stroke-dashoffset: 63.460px;"
        />
      </svg>
    </span>
    <div
      class="MuiBox-root css-1o8dslu"
    >
      <div
        class="MuiTypography-root MuiTypography-h4 css-1pxewg2-MuiTypography-root"
      >
        50%
      </div>
    </div>
  </div>
</div>
`;
